Close (#22) Created Obesity ADVS Vignette (#26)
pharmaverse/admiralmetabolic: vignettes/advs.Rmd

We advise you first consult the `{admiral}` [Creating a BDS Finding ADaM
vignette](https://pharmaverse.github.io/admiral/articles/bds_finding.html).
The programming workflow around creating the general set-up of an ADVS using
`{admiral}` functions is the same. We focus on the most common endpoints and their derivations mainly found in metabolic trials to avoid repeating information and maintaining the same content in two places.

# Programming Workflow

* [Initial ADVS Set-up](#advs_start)
* [Derive Additional Parameters (e.g. `BMI` for `ADVS`)](#derive_param)
* [Derive Categorization Variables (`AVALCATx`, `BASECATx`)](#cat)
* [Derive Criterion Variables (`CRITy`, `CRITyFL`, `CRITyFLN`)](#crit_vars)
* [Remaining ADVS Set-up](#advs_end)

## Initial ADVS Set-up {#advs_start}

The following steps are to merge ADSL variables with the source data and derive 
the usual ADVS analysis variables. The `{admiral}` [Creating a BDS Finding ADaM
vignette](https://pharmaverse.github.io/admiral/articles/bds_finding.html)
gives detailed guidance on all these steps.

The only difference here would be the the additional parameters collected in 
metabolic trials - for example:

> `VSTESTCD` | `PARAMCD` | `PARAM` | `PARAMN` | `PARCAT1` | `PARCAT1N`

Author: > Wondering whether it would make more sense to showcase this table as an R tibble, so that users can copy it if they wish? 
Agreed. I'll add a tibble instead of this table. 

As it stands I can't easily replicate the mappings shown here.
I didn't understand this statement

--------- | --------- | -------- | ------- | --------- | ----------
BMI | BMI | Body Mass Index (kg/m^2) | 1 | Subject Characteristic | 1
HIPCIR | HIPCIR | Hip Circumference (cm) | 3 | Subject Characteristic | 1
WSTCIR | WSTCIR | Waist Circumference (cm) | 5 | Subject Characteristic | 1

> ```{r, message=FALSE, include=FALSE}

Reviewer: Move some of this to the "read in data" section

advs <-
  derive_vars_dy(advs, reference_date = TRTSDT, source_vars = exprs(ADT))

param_lookup <- tribble(
  ~VSTESTCD, ~PARAMCD, ~PARAM, ~PARAMN,
  "HEIGHT", "HEIGHT", "Height (cm)", 2,
  "HIPCIR", "HIPCIR", "Hip Circumference (cm)", 3,
  "WEIGHT", "WEIGHT", "Weight (kg)", 4,
  "WSTCIR", "WSTCIR", "Waist Circumference (cm)", 5
)
attr(param_lookup$VSTESTCD, "label") <- "Vital Signs Test Short Name"

advs <- derive_vars_merged_lookup(
  advs,
  dataset_add = param_lookup,
  new_vars = exprs(PARAMCD),
  by_vars = exprs(VSTESTCD)
)

advs <- mutate(
  advs,
  AVAL = VSSTRESN
)
```

## Derive Additional Parameters (e.g. `BMI` for `ADVS`) {#derive_param}

In metabolic trials, `BMI` is often calculated at source using collected height and weight for screening purposes. But while creating the `ADVS` dataset, we re-derive `BMI` from the collected height and weight values. This is done to ensure that the `BMI` is calculated consistently across all subjects and visits. 

In this step, we create parameter `Body Mass Index` (BMI) for `ADVS` domain using the `derive_param_bmi()` function.
Note that only variables specified in the `by_vars` argument will be populated in the newly 

> created records. Also note that if height is collected only once use `constant_by_vars` to specify the subject-level variable to merge on.  Otherwise BMI is only calculated for visits where both are collected.

Reviewer: "Also note that if height is collected only once use `constant_by_vars` to specify the subject-level variable to merge on" - sentence reads a bit weirdly, please rephrase

Author: "Also note that if height is collected only once for a subject use constant_by_vars to specify the function to merge by the subject-level variable" 
How does this look ? The original sentence comes from ADVS BDS vignette.

```{r eval=TRUE}
# Removing BMI collected at source from the dataset
advs <- advs %>% filter(!VSTESTCD == "BMI")

advs <- derive_param_bmi(
  advs,

>   by_vars = exprs(STUDYID, USUBJID, !!!adsl_vars, VISIT, VISITNUM, ADT, ADY, VSTPT, VSTPTNUM),

Reviewer: Use `get_admiral_option("subject_keys")` rather than explicitly naming STUDYID and USUBJID - see how it's done in [this vignette](https://pharmaverse.github.io/admiralophtha/articles/advfq.html) for an example

  set_values_to = exprs(
    PARAMCD = "BMI",
    PARAM = "Body Mass Index (kg/m^2)",
    PARAMN = 1
  ),
  get_unit_expr = VSSTRESU,
  filter = VSSTAT != "NOT DONE" | is.na(VSSTAT),
  constant_by_vars = exprs(USUBJID)
)
```

```{r, eval=TRUE, echo=FALSE}
dataset_vignette(
  arrange(advs, USUBJID, VISITNUM, VSTPTNUM, ADT, PARAMCD),
  display_vars = exprs(USUBJID, VSTESTCD, PARAMCD, PARAM, VISIT, AVAL),
  filter = PARAMCD %in% c("BMI")
)
```

```{r eval=TRUE, include=FALSE}
advs <- restrict_derivation(
  advs,
  derivation = derive_var_extreme_flag,
  args = params(
    by_vars = exprs(STUDYID, USUBJID, PARAMCD),
    order = exprs(ADT, VSTPTNUM, VISITNUM),
    new_var = ABLFL,
    mode = "last"
  ),
  filter = (!is.na(AVAL) & ADT <= TRTSDT)
)

advs <- derive_var_base(
  advs,
  by_vars = exprs(STUDYID, USUBJID, PARAMCD),
  source_var = AVAL,
  new_var = BASE
)

advs <- derive_var_chg(advs)

advs <- derive_var_pchg(advs)
```

> 

Reviewer: I recommend adding a section here explaining what the common endpoints in an obesity trail are, and how we recommend coding them into ADVS. This will help justify the `AVALCAT` and `CRITFL` sections that come after.

## Derive Categorization Variables (`AVALCATx`, `BASECATx`) {#cat}

> Admiral does not currently have a generic function to aid in assigning `AVALCATy`/

Reviewer: ```suggestion
`{admiral}` does not currently have a generic function to aid in assigning `AVALCATy`/
```

> Admiral does not currently have a generic function to aid in assigning `AVALCATy`/

Reviewer: While this is true right now, take a look [here](https://github.com/pharmaverse/admiral/issues/2480) as it's coming! I don't know when the [corresponding PR](https://github.com/pharmaverse/admiral/pull/2503) will be merged, but when it is, we should update this section to use the function instead. Maybe @Siddhesh2097 you can make an issue for it in admiralmetabolic in preparation?

  ~PARAMCD, ~BASECA1N, ~BASECAT1,
  "BMI", 1, "Underweight",
  "BMI", 2, "Normal weight",
  "BMI", 3, "Overweight",
  "BMI", 4, "Obesity class I",
  "BMI", 5, "Obesity class II",
  "BMI", 6, "Obesity class III",
  "BMI", NA, NA_character_
)

format_basecat1n <- function(param, base) {
  case_when(
    param == "BMI" & base < 18.5 ~ 1,
    param == "BMI" & base >= 18.5 & base < 25 ~ 2,
    param == "BMI" & base >= 25 & base < 30 ~ 3,
    param == "BMI" & base >= 30 & base < 35 ~ 4,
    param == "BMI" & base >= 35 & base < 40 ~ 5,
    param == "BMI" & base >= 40 ~ 6,
    TRUE ~ NA_real_
  )
}
advs <- advs %>%
  mutate(BASECA1N = format_basecat1n(param = PARAMCD, base = BASE)) %>%
  derive_vars_merged(
    basecat_lookup,
    by = exprs(PARAMCD, BASECA1N)
  )
```

```{r, eval=TRUE, echo=FALSE}
dataset_vignette(
  advs,
  display_vars = exprs(USUBJID, PARAMCD, VISIT, AVAL, BASE, ABLFL, BASECA1N, BASECAT1),
  filter = PARAMCD == "BMI"
)
```

## Derive Criterion Variables (`CRITy`, `CRITyFL`, `CRITyFLN`) {#crit_vars}

For deriving criterion variables (`CRITy`, `CRITyFL`, `CRITyFLN`) `{admiral}`
provides `derive_vars_crit_flag()`. It ensures that they are derived in an

> ADaM-compliant way (see documentation of the function for details).

Reviewer: link the function documentation from the admiral website

In most cases the criterion depends on the parameter. In the
following example, the criterion flags for weight based on percentage change in 
weight reduction is derived. Additional criterion flags can be added as needed.

```{r eval=TRUE}
advs <- advs %>%
  derive_vars_crit_flag(
    condition = PCHG <= -5 & PARAMCD == "WEIGHT",

>     description = "Achievement of ≥ 5% weight reduction",

Reviewer: should this say - from baseline? and similar below

endpoints and their derivations mainly found in metabolic trials to
avoid repeating information and maintaining the same content in two
places.

# Programming Workflow

-   [Read in Data](#readdata)
-   [Initial ADVS Set-up](#advs_start)
-   [Derive Additional Parameters (e.g. `BMI` for `ADVS`)](#derive_param)
-   [Derive Categorization Variables (`AVALCATx`, `BASECATx`)](#cat)
-   [Derive Criterion Variables (`CRITy`, `CRITyFL`,
    `CRITyFLN`)](#crit_vars)
-   [Remaining ADVS Set-up](#advs_end)

## Read in Data {#readdata}

To start, all data frames needed for the creation of `ADVS` should be
read into the environment. This will be a company specific process. Some
of the data frames needed may be `VS` and `ADSL`.

```{r, message=FALSE, warning=FALSE}
library(admiral)
library(admiralmetabolic)
library(pharmaversesdtm)
library(dplyr, warn.conflicts = FALSE)
library(lubridate)
library(stringr)

data("dm_metabolic")
data("vs_metabolic")

dm <- convert_blanks_to_na(dm_metabolic)
vs <- convert_blanks_to_na(vs_metabolic)
```

```{r, include=FALSE }
data("admiral_adsl")

dm <- dm %>%
  left_join(admiral_adsl %>% select(USUBJID, TRTSDT, TRTEDT), by = "USUBJID")
```

> The `DM` domain is used as the basis for `ADSL`:

Reviewer: ```suggestion
Within this vignette, the `DM` domain is used as the basis for `ADSL`:
```

> ```

Reviewer 2: This probably needs to be included.

```{r eval=TRUE}
adsl <- dm %>%
  select(-DOMAIN)
```

> ```{r, eval=TRUE, echo=FALSE}

Reviewer: Do we need to display DM here? I'd cut out this dataset vignette.

The following steps are to merge ADSL variables with the source data and
derive the usual ADVS analysis variables. The `{admiral}` [Creating a
BDS Finding ADaM
vignette](https://pharmaverse.github.io/admiral/articles/bds_finding.html)
gives detailed guidance on all these steps.

> ```{r, include=FALSE}

Reviewer 2: Should be included.

adsl <- adsl %>%
  mutate(
    TRT01P = ARM,
    TRT01A = ACTARM
  )

adsl_vars <- exprs(TRTSDT, TRTEDT, TRT01P, TRT01A)

advs <- derive_vars_merged(
  vs,
  dataset_add = adsl,
  new_vars = adsl_vars,
  by_vars = exprs(STUDYID, USUBJID)
)

advs <- derive_vars_dt(advs, new_vars_prefix = "A", dtc = VSDTC)

advs <-
  derive_vars_dy(advs, reference_date = TRTSDT, source_vars = exprs(ADT))
```

The only difference here would be the the additional parameters
collected in metabolic trials - `BMI`, `HIPCIR` and `WSTCIR`. Next step
is to assign parameter level values such as `PARAMCD`, `PARAM`,
`PARAMN`, `PARCAT1`, etc.,and for this a lookup can be created based on
the SDTM `--TESTCD` value to join to the source data.

```{r, include=FALSE}
# `VSTESTCD` | `PARAMCD` | `PARAM` | `PARAMN` | `PARCAT1` | `PARCAT1N`
# --------- | --------- | -------- | ------- | --------- | ----------
# BMI | BMI | Body Mass Index (kg/m^2) | 1 | Subject Characteristic | 1
# HIPCIR | HIPCIR | Hip Circumference (cm) | 3 | Subject Characteristic | 1
# WSTCIR | WSTCIR | Waist Circumference (cm) | 5 | Subject Characteristic | 1

> ```

Reviewer 2: It appears that this code chunck does not do anything

# `VSTESTCD` | `PARAMCD` | `PARAM` | `PARAMN` | `PARCAT1` | `PARCAT1N`
# --------- | --------- | -------- | ------- | --------- | ----------
# BMI | BMI | Body Mass Index (kg/m^2) | 1 | Subject Characteristic | 1
# HIPCIR | HIPCIR | Hip Circumference (cm) | 3 | Subject Characteristic | 1
# WSTCIR | WSTCIR | Waist Circumference (cm) | 5 | Subject Characteristic | 1
```

```{r, echo=TRUE, message=FALSE}

> param_lookup <- tribble(

Reviewer: This looks good now - only ask is that you right-align the table - [see here](https://pharmaverse.github.io/admiral/dev/articles/visits_periods.html#creating-the-periodphase-reference-dataset-from-ex) for an example.

Author: @manciniedoardo Styler throws an error with right aligned lookup table. Have kept the original one for now.

  ~VSTESTCD, ~PARAMCD, ~PARAM, ~PARAMN, ~PARCAT1, ~PARCAT1N,
  "HEIGHT", "HEIGHT", "Height (cm)", 1, "Subject Characteristic", 1,
  "WEIGHT", "WEIGHT", "Weight (kg)", 2, "Subject Characteristic", 1,
  "BMI", "Body Mass Index", "Body Mass Index(kg/m^2)", 3, "Subject Characteristic", 1,
  "HIPCIR", "HIPCIR", "Hip Circumference (cm)", 4, "Subject Characteristic", 1,
  "WSTCIR", "WSTCIR", "Waist Circumference (cm)", 5, "Subject Characteristic", 1,
  "DIABP", "DIABP", "Diastolic Blood Pressure (mmHg)", 6, "Vital Sign", 2,
  "PULSE", "PULSE", "Pulse Rate (beats/min)", 7, "Vital Sign", 2,
  "SYSBP", "SYSBP", "Systolic Blood Pressure (mmHg)", 8, "Vital Sign", 2,
  "TEMP", "TEMP", "Temperature (C)", 9, "Vital Sign", 2
)
attr(param_lookup$VSTESTCD, "label") <- "Vital Signs Test Short Name"
```

> This lookup may now be joined to the source data and this is how the parameters will look like.

Reviewer: ```suggestion
This lookup may now be joined to the source data and this is how the parameters will look like:
```

> ```{r, eval=TRUE, include=FALSE}

Reviewer: I would include this call:

```suggestion
```{r, eval=TRUE, include=TRUE}
```

  by_vars = exprs(VSTESTCD)
)
```

```{r, eval=TRUE, echo=FALSE}
advs_param <- distinct(advs, USUBJID, PARAMCD, VSTESTCD, PARAM, PARCAT1, PARCAT1N)

dataset_vignette(advs_param, display_vars = exprs(USUBJID, VSTESTCD, PARAMCD, PARAM, PARCAT1, PARCAT1N))
```

```{r, eval=TRUE, include=FALSE}
advs <- mutate(
  advs,
  AVAL = VSSTRESN
)

> ```

Reviewer 2: Probably needs to be included. The next code chunck does not run without it.

## Derive Additional Parameters (e.g. `BMI` for `ADVS`) {#derive_param}

In metabolic trials, `BMI` is often calculated at source using collected
height and weight for screening purposes. But while creating the `ADVS`
dataset, we re-derive `BMI` from the collected height and weight values.
This is done to ensure that the `BMI` is calculated consistently across
all subjects and visits.

> In this step, we create parameter `Body Mass Index` (BMI) for the `ADVS`

Reviewer: ```suggestion
In this step, we create parameter Body Mass Index (BMI) for the `ADVS`
```

> In this step, we create parameter `Body Mass Index` (BMI) for `ADVS`

Reviewer: ```suggestion
In this step, we create parameter `Body Mass Index` (BMI) for the `ADVS`
```

domain using the `derive_param_bmi()` function. Note that only variables
specified in the `by_vars` argument will be populated in the newly

> created records. "Also note that if height is collected only once for a

Reviewer: ```suggestion
created records. Also note that if height is collected only once for a
```

subject use `constant_by_vars` to specify the function to merge by the

> subject-level variable" Otherwise BMI is only calculated for visits

Reviewer: ```suggestion
subject-level variable - otherwise BMI is only calculated for visits
```

advs <- derive_param_bmi(
  advs,
  by_vars = c(
    get_admiral_option("subject_keys"),
    exprs(!!!adsl_vars, VISIT, VISITNUM, ADT, ADY, VSTPT, VSTPTNUM)
  ),
  set_values_to = exprs(
    PARAMCD = "BMI",
    PARAM = "Body Mass Index (kg/m^2)",
    PARAMN = 3,
    PARCAT1 = "Subject Characteristic",
    PARCAT1N = 1
  ),
  get_unit_expr = VSSTRESU,
  filter = VSSTAT != "NOT DONE" | is.na(VSSTAT),
  constant_by_vars = exprs(USUBJID)
)
```

```{r, eval=TRUE, echo=FALSE}
dataset_vignette(
  arrange(advs, USUBJID, VISITNUM, VSTPTNUM, ADT, PARAMCD),
  display_vars = exprs(USUBJID, VSTESTCD, PARAMCD, PARAM, VISIT, AVAL),
  filter = PARAMCD %in% c("BMI")
)
```

```{r eval=TRUE, include=FALSE}
advs <- restrict_derivation(
  advs,
  derivation = derive_var_extreme_flag,
  args = params(
    by_vars = c(get_admiral_option("subject_keys"), exprs(PARAMCD)),
    order = exprs(ADT, VSTPTNUM, VISITNUM),
    new_var = ABLFL,
    mode = "last"
  ),
  filter = (!is.na(AVAL) & ADT <= TRTSDT)
)

advs <- derive_var_base(
  advs,
  by_vars = c(get_admiral_option("subject_keys"), exprs(PARAMCD)),
  source_var = AVAL,
  new_var = BASE
)

advs <- derive_var_chg(advs)

advs <- derive_var_pchg(advs)
```

> In the following sections, we will explore some of the most common

Reviewer: I think this would read better if you add a new section header here (maybe "Common Metabolism Endpoints") and make the next two sections into subsections.

endpoints typically observed in metabolic trials.

One such endpoint is the improvement in weight category from baseline to
the end of treatment, which is often assessed using Body Mass Index
(`BMI`). To capture this, we will derive variables such as `AVALCATy`
and `BASECATy`, as detailed in the subsequent section.

Additionally, the achievement of weight reduction thresholds, such as ≥

> 5%, ≥ 10%, or ≥ 15% from baseline to end of treatment or certain visit,

Reviewer: ```suggestion
5%, ≥ 10%, or ≥ 15% from baseline to end of treatment or a certain visit,
```

> is a common endpoint in obesity trials. To accommodate these criteria,

Reviewer: Are we referring to metabolic or obesity trials? Are they synonymous? Maybe we could standardise the language?

we will derive relevant criterion variables such as `CRITy`, `CRITyFL`,
and `CRITyFLN`, with the necessary functions provided by `{admiral}`
outlined below.

> #### Derive Categorization Variables (`AVALCATx`, `BASECATx`) {#cat}

Reviewer: ```suggestion
### Derive Categorization Variables (`AVALCATx`, `BASECATx`) {#cat}
```

adsl <- dm %>%
  select(-DOMAIN) %>%
  mutate(TRT01P = ARM, TRT01A = ACTARM) %>%
  left_join(admiral_adsl %>% select(USUBJID, TRTSDT, TRTEDT), by = "USUBJID")
```

The following steps are to merge `ADSL` variables with the source data and
derive the usual `ADVS` analysis variables. Note that only the sections required for this vignette are covered in the following steps. To get a detailed guidance on all the steps, refer the `{admiral}` [Creating a BDS Finding ADaM vignette](https://pharmaverse.github.io/admiral/articles/bds_finding.html).

Join `ADSL` to your `VS` domain. Only the `ADSL` variables used for derivations are selected at this step.
```{r, eval=TRUE}
adsl_vars <- exprs(TRTSDT, TRTEDT, TRT01P, TRT01A)

advs <- derive_vars_merged(
  vs,
  dataset_add = adsl,
  new_vars = adsl_vars,
  by_vars = exprs(STUDYID, USUBJID)
)
```

The function `derive_vars_dt()` can be used to derive `ADT`.

```{r, eval=TRUE}
advs <- derive_vars_dt(advs, new_vars_prefix = "A", dtc = VSDTC)
```

Once `ADT` is derived, the function `derive_vars_dy()` can be used to derive `ADY`

```{r, eval=TRUE}
advs <-
  derive_vars_dy(advs, reference_date = TRTSDT, source_vars = exprs(ADT))

> ```

Author: Hiding this

## Assign `PARAMCD`, `PARAM`, `PARAMN`, `PARCAT1` {#paramcd}

The next step is to assign parameter level values such as `PARAMCD`, `PARAM`,
`PARAMN`, `PARCAT1`, etc. For this, a lookup can be created based on
the SDTM `--TESTCD` value to join to the source data.

> The only difference here would be the the additional parameters
> collected in metabolic trials - `BMI`, `HIPCIR` and `WSTCIR`.

Author: ```suggestion
One key distinction to note is the additional parameters collected in metabolic trials - `BMI`, `HIPCIR`, and `WSTCIR`.
```
The idea was to highlight the additional parameters collected in this vignette that are not there in standard one and display them in this section. To not cause confusion , can we rephrase it as suggested above ?
@AndersAskeland

---
title: "Creating a Metabolic ADVS ADaM"
output: 
  rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Creating a Metabolic ADVS ADaM}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

library(admiraldev)
```

# Introduction

This article describes creating a vital signs ADaM for metabolic
clinical trials.

We advise you first consult the `{admiral}` [Creating a BDS Finding ADaM
vignette](https://pharmaverse.github.io/admiral/articles/bds_finding.html).
The programming workflow around creating the general set-up of an `ADVS`
using `{admiral}` functions is the same. We focus on the most common
endpoints and their derivations mainly found in metabolic trials to
avoid repeating information and maintaining the same content in two

> places.

Reviewer: ```suggestion
places. As such, the code in this vignette is not completely executable; we recommend consulting the ADVS template script to view the full workflow.
```

## Required Packages

The examples of this vignette require the following packages.

```{r, warning=FALSE, message=FALSE}
library(admiral)
library(admiralmetabolic)
library(pharmaversesdtm)

> library(dplyr, warn.conflicts = FALSE)

Reviewer: I think it's because otherwise the message conflicts can get rendered in the vignette, which is quite ugly to read

> library(dplyr, warn.conflicts = FALSE)

Reviewer 2: I agree that would be ugly, however, as we are using `message = FALSE`, I would suggest we remove the `warn.conflicts` argument.

library(lubridate)
library(stringr)
```

# Programming Workflow

-   [Read in Data](#readdata)
-   [Assign `PARAMCD`, `PARAM`, `PARAMN`, `PARCAT1`](#paramcd)
-   [Derive Additional Parameters (e.g. `BMI` for `ADVS`)](#derive_param)
-   [Common Metabolic Endpoints](#common_endpoints)
    -   [Derive Categorization Variables (`AVALCATx`, `BASECATx`)](#cat)
    -   [Derive Criterion Variables (`CRITy`, `CRITyFL`, `CRITyFLN`)](#crit_vars)
-   [Remaining ADVS Set-up](#advs_end)

## Read in Data {#readdata}

To start, all data frames needed for the creation of `ADVS` should be
read into the environment. This will be a company specific process. Some

> of the data frames needed may be `VS` and `ADSL`.

Reviewer: FYI i updated this comment to say pharmaversesdtm rather than admiralsdtm

```{r, message=FALSE, warning=FALSE}
dm_metabolic <- admiralmetabolic::dm_metabolic
vs_metabolic <- admiralmetabolic::vs_metabolic
admiral_adsl <- admiral::admiral_adsl

dm <- convert_blanks_to_na(dm_metabolic)
vs <- convert_blanks_to_na(vs_metabolic)
admiral_adsl <- convert_blanks_to_na(admiral_adsl)
```

Within this vignette, the `DM` is used as the basis for `ADSL`:

```{r eval=TRUE}
# Retrieve required variables from admiral ADSL for this vignette that are not present in DM dataset
adsl <- dm %>%
  select(-DOMAIN) %>%
  mutate(TRT01P = ARM, TRT01A = ACTARM) %>%
  left_join(admiral_adsl %>% select(USUBJID, TRTSDT, TRTEDT), by = "USUBJID")
```

The following steps are to merge `ADSL` variables with the source data and
derive the usual `ADVS` analysis variables. Note that only the sections required for this vignette are covered in the following steps. To get a detailed guidance on all the steps, refer the `{admiral}` [Creating a BDS Finding ADaM vignette](https://pharmaverse.github.io/admiral/articles/bds_finding.html).

> Join `ADSL` to your `VS` domain. Only the `ADSL` variables used for derivations are selected at this step.

Reviewer: From here to the end of the subsection, I would hide these chunks, remove the explanation and accept the vignette is not completely executable as otherwise we are just duplicating what is in the admiral advs vignette.

Author: @manciniedoardo  A bit confused here, are we not showing the next step of creating advs as you have marked here ?

Reviewer: I just think we should hide the bits where you've just copied from the ADVS vignette (deriving dates, days, merging with adsl)

```{r, eval=TRUE}
adsl_vars <- exprs(TRTSDT, TRTEDT, TRT01P, TRT01A)

advs <- derive_vars_merged(
  vs,
  dataset_add = adsl,
  new_vars = adsl_vars,
  by_vars = exprs(STUDYID, USUBJID)
)
```

The function `derive_vars_dt()` can be used to derive `ADT`.

```{r, eval=TRUE}
advs <- derive_vars_dt(advs, new_vars_prefix = "A", dtc = VSDTC)
```

Once `ADT` is derived, the function `derive_vars_dy()` can be used to derive `ADY`

```{r, eval=TRUE}
advs <-
  derive_vars_dy(advs, reference_date = TRTSDT, source_vars = exprs(ADT))
```

## Assign `PARAMCD`, `PARAM`, `PARAMN`, `PARCAT1` {#paramcd}

Next step is to assign parameter level values such as `PARAMCD`, `PARAM`,
`PARAMN`, `PARCAT1`, etc.,and for this a lookup can be created based on
the SDTM `--TESTCD` value to join to the source data.
The only difference here would be the the additional parameters

> collected in metabolic trials - `BMI`, `HIPCIR` and `WSTCIR`.

Reviewer 2: Do you mean only difference compared to the standard BDS vignette? I would perhaps omit this sentence, as it could cause some confusion.

Author: ```suggestion
The only difference here would be the the additional parameters
collected in metabolic trials - `BMI`, `HIPCIR` and `WSTCIR`.
```

The idea was to highlight the additional parameters collected in this vignette that are not there in standard one and display them in this section. To not cause confusion , can we rephrase it as suggested above ?
@AndersAskeland

Reviewer 2: I think my main thought here, is that it some somewhat ambiguous to what the "key distinction"/"difference" is in relation to. I believe it is compared to the standard Admiral BDS vignette, but it is not entirely clear. Perhaps one can write something similar to:

```suggestion
One key addition in metabolic trials are vital sign parameters associated to body measurements, such as `BMI`, `HIPCIR`, and `WSTCIR`.
```

```{r, echo=TRUE, message=FALSE}
param_lookup <- tribble(
  ~VSTESTCD, ~PARAMCD, ~PARAM, ~PARAMN, ~PARCAT1, ~PARCAT1N,
  "HEIGHT", "HEIGHT", "Height (cm)", 1, "Subject Characteristic", 1,
  "WEIGHT", "WEIGHT", "Weight (kg)", 2, "Subject Characteristic", 1,
  "BMI", "BMI", "Body Mass Index(kg/m^2)", 3, "Subject Characteristic", 1,
  "HIPCIR", "HIPCIR", "Hip Circumference (cm)", 4, "Subject Characteristic", 1,
  "WSTCIR", "WSTCIR", "Waist Circumference (cm)", 5, "Subject Characteristic", 1,
  "DIABP", "DIABP", "Diastolic Blood Pressure (mmHg)", 6, "Vital Sign", 2,
  "PULSE", "PULSE", "Pulse Rate (beats/min)", 7, "Vital Sign", 2,
  "SYSBP", "SYSBP", "Systolic Blood Pressure (mmHg)", 8, "Vital Sign", 2,
  "TEMP", "TEMP", "Temperature (C)", 9, "Vital Sign", 2
)

> attr(param_lookup$VSTESTCD, "label") <- "Vital Signs Test Short Name"

Reviewer: No need to label here

```

This lookup may now be joined to the source data and this is how the
parameters will look like:

```{r, eval=TRUE, include=TRUE, message=FALSE}
advs <- derive_vars_merged_lookup(
  advs,
  dataset_add = param_lookup,
  new_vars = exprs(PARAMCD, PARAM, PARAMN, PARCAT1, PARCAT1N),
  by_vars = exprs(VSTESTCD)
)
```

```{r, eval=TRUE, echo=FALSE}
advs_param <- distinct(advs, USUBJID, PARAMCD, VSTESTCD, PARAM, PARCAT1, PARCAT1N)

dataset_vignette(advs_param, display_vars = exprs(USUBJID, VSTESTCD, PARAMCD, PARAM, PARCAT1, PARCAT1N))
```

The mapping of `AVAL` is left to the ADaM programmer. An example mapping may be:

```{r, eval=TRUE}
advs <- mutate(
  advs,
  AVAL = VSSTRESN
)
```

## Derive Additional Parameters (e.g. `BMI` for `ADVS`) {#derive_param}

> In metabolic trials, `BMI` is often calculated at source using collected
> height and weight for screening purposes. But while creating the `ADVS`

Reviewer 2: ```suggestion
In metabolic trials, `BMI` is often calculated at source. But while creating the `ADVS`
```

dataset, we re-derive `BMI` from the collected height and weight values.
This is done to ensure that the `BMI` is calculated consistently across
all subjects and visits.

In this step, we create parameter Body Mass Index (`BMI`) for the `ADVS`
domain using the `derive_param_bmi()` function. Note that only variables
specified in the `by_vars` argument will be populated in the newly
created records. Also note that if height is collected only once for a
subject use `constant_by_vars` to specify the function to merge by the
subject-level variable - otherwise BMI is only calculated for visits
where both are collected.

```{r eval=TRUE}
# Removing BMI collected at source from the dataset
advs <- advs %>% filter(!VSTESTCD == "BMI")

advs <- derive_param_bmi(
  advs,
  by_vars = c(
    get_admiral_option("subject_keys"),
    exprs(!!!adsl_vars, VISIT, VISITNUM, ADT, ADY, VSTPT, VSTPTNUM)
  ),
  set_values_to = exprs(
    PARAMCD = "BMI",
    PARAM = "Body Mass Index (kg/m^2)",

>     PARAMN = 3,

Reviewer 3: It seems that in this step PARAMN for BMI is set, but at the same time PARAMN for all the other parameters is set to NA, i.e. the PARAMN values that were previously there for the other parameters do not remain.

Reviewer 4: Are you sure? This would be a bug in the admiral function. The values specified for `set_values_to` should be used for the new records only.

From the code it seems that `PARAMN` is not derived before. It is included in the lookup table but not merged to the data (`PARAMN` is not included in `new_vars` in the `derive_vars_merged_lookup()` call).

Reviewer 3: You are right. Thanks @bundfussr.
I had seen the `PARAMN` only in the lookup table but not in the ADVS dataset itself. I will just make a suggestion to add `PARAMN` for the other parameters so that not only BMI get's a `PARAMN`.

function `derive_var_extreme_flag()`.  For example, if baseline is defined as the last 
non-missing `AVAL` prior or on `TRTSDT`, the function call for `ABLFL` would be:

```{r eval=TRUE}
advs <- restrict_derivation(
  advs,
  derivation = derive_var_extreme_flag,
  args = params(
    by_vars = c(get_admiral_option("subject_keys"), exprs(PARAMCD)),
    order = exprs(ADT, VSTPTNUM, VISITNUM),
    new_var = ABLFL,
    mode = "last"
  ),
  filter = (!is.na(AVAL) & ADT <= TRTSDT)
)
```

Lastly, the `BASE` can be derived using the `{admiral}` function
`derive_var_base()`. Example:

```{r eval=TRUE}
advs <- derive_var_base(
  advs,
  by_vars = c(get_admiral_option("subject_keys"), exprs(PARAMCD)),
  source_var = AVAL,
  new_var = BASE
)
```

In this step, we derive change and percent change from baseline using the `{admiral}` 
functions `derive_var_chg()` and `derive_var_pchg()`.  These functions expect `AVAL` 
and `BASE` to exist in the data frame.  The `CHG` is simply `AVAL - BASE` and the
`PCHG` is `(AVAL - BASE) / absolute value (BASE) * 100`.  Example calls are:

```{r eval=TRUE}
advs <- derive_var_chg(advs)

> 

Reviewer: Remove empty line here

advs <- derive_var_pchg(advs)
```

## Common Metabolic Endpoints {#common_endpoints}

In the following sections, we will explore some of the most common
endpoints typically observed in metabolic trials.

One such endpoint is the improvement in weight category from baseline to
the end of treatment, which is often assessed using Body Mass Index
(`BMI`). To capture this, we will derive variables such as `AVALCATy`
and `BASECATy`, as detailed in the subsequent section.

Additionally, the achievement of weight reduction thresholds, such as ≥

> 5%, ≥ 10%, or ≥ 15% from baseline to end of treatment or a certain visit,

Reviewer 2: ```suggestion
5%, ≥ 10%, or ≥ 15% from baseline to end of treatment or at a certain visit,
```

assigning `AVALCATy`/ `AVALCAvN` and `BASECATy`/ `BASECAvN` values.
Below is a simple example of how these values may be assigned:

```{r eval=TRUE}
avalcat_lookup <- tibble::tribble(
  ~PARAMCD, ~AVALCA1N, ~AVALCAT1,
  "BMI",    1,         "Underweight",
  "BMI",    2,         "Normal weight",
  "BMI",    3,         "Overweight",
  "BMI",    4,         "Obesity class I",
  "BMI",    5,         "Obesity class II",
  "BMI",    6,         "Obesity class III",
  "BMI",    NA,        NA_character_
)

format_avalcat1n <- function(param, aval) {
  case_when(
    param == "BMI" & aval < 18.5 ~ 1,
    param == "BMI" & aval >= 18.5 & aval < 25 ~ 2,
    param == "BMI" & aval >= 25 & aval < 30 ~ 3,
    param == "BMI" & aval >= 30 & aval < 35 ~ 4,
    param == "BMI" & aval >= 35 & aval < 40 ~ 5,
    param == "BMI" & aval >= 40 ~ 6,
    TRUE ~ NA_real_
  )
}

advs <- advs %>%
  mutate(AVALCA1N = format_avalcat1n(param = PARAMCD, aval = AVAL)) %>%
  derive_vars_merged(
    avalcat_lookup,
    by = exprs(PARAMCD, AVALCA1N)
  )
```

```{r eval=TRUE, echo=FALSE}
dataset_vignette(
  advs,
  display_vars = exprs(USUBJID, PARAMCD, VISIT, AVAL, AVALCA1N, AVALCAT1),
  filter = PARAMCD == "BMI"
)
```

> In the similar way, we will create `BASECATy`/ `BASECAvN` variables.

Reviewer 2: ```suggestion
In a similar way, we will create `BASECATy`/ `BASECAvN` variables.
```
